corehq/apps/cloudcare/models.py

```diff
@@ -68,34 +87,59 @@ def get_by_domain(cls, domain):
         return self or cls(domain=domain)
 
     def user_can_access_app(self, user, app):
-        user_id = user['_id']
         app_id = app['_id']
-        if not self.restrict or user['doc_type'] == 'WebUser':
+        app_ids = (app_id, app['copy_of'] or ())
+
+        if not self.restrict or isinstance(user, (WebUser, OTARestoreWebUser)):
             return True
+
+        return self.group_permitted(app_ids, user) or self.userdata_permitted(app_ids, user)
+
+    def group_permitted(self, app_ids, user):
         app_group = None
         for app_group in self.app_groups:
-            if app_group.app_id in (app_id, app['copy_of'] or ()):
+            if app_group.app_id in app_ids:
                 break
         if app_group:
```

Comment:
Yeah, this is a really weird construct. I didn't actually write this, and it's been in prod since 2012, so I left it as is. But yes..